File: scala/aoc/src/main/scala/adventofcode/Day17/Day17.scala
package adventofcode.Day17

import scala.annotation.tailrec

case class Target(xMin: Int, xMax: Int, yMin: Int, yMax: Int)

case class State(xPos: Int, yPos: Int, xVelocity: Int, yVelocity: Int):
  def hitTarget(target: Target): Boolean =
    xPos >= target.xMin && xPos <= target.xMax && yPos >= target.yMin && yPos <= target.yMax

  def overshot(target: Target): Boolean =
    xPos > target.xMax || yPos < target.yMin

  def update: State =
    val updatedVelocity =
      if xVelocity < 0 then xVelocity + 1
      else if xVelocity > 0 then xVelocity - 1
      else 0
    State(xPos + xVelocity, yPos + yVelocity, updatedVelocity, yVelocity - 1)

def runSimulation(initialState: State, target: Target): Option[Int] =
  @tailrec
  def loop(state: State, yMax: Int, n: Int): Option[Int] =
    if n > 1000 then None
    else if state.overshot(target) then None
    else if state.hitTarget(target) then Some(yMax)
    else
      val newState = state.update
      loop(newState, scala.math.max(yMax, newState.yPos), n + 1)

  loop(initialState, Integer.MIN_VALUE, 0)


@main def day17(): Unit =
  val target = Target(85, 145, -163, -108)
  val heightsOption = for xV <- Range(0, 200)
                          yV <- Range(-200, 200)
  yield runSimulation(State(0, 0, xV, yV), target)
  val heights = heightsOption.flatten

  val part1 = heights.max
  val part2 = heights.length

  println(part1)
  println(part2)

File: clojure/aoc/src/aoc/day17.clj
(ns aoc.day17)

(defrecord State [x y x-velocity y-velocity])
(defrecord Target [x-min x-max y-min y-max])

(defn hit-target? [target state]
    (and (>= (:x state) (:x-min target))
         (<= (:x state) (:x-max target))
         (>= (:y state) (:y-min target))
         (<= (:y state) (:y-max target))))

(defn overshot? [target state]
  (or (> (:x state) (:x-max target))
      (< (:y state) (:y-min target))))

(defn update-state [state]
  (->State
    (+ (:x state) (:x-velocity state))
    (+ (:y state) (:y-velocity state))
    (cond
      (< (:x-velocity state) 0) (inc (:x-velocity state))
      (> (:x-velocity state) 0) (dec (:x-velocity state))
      :else 0)
    (dec (:y-velocity state))))

(defn run-simulation [x-v y-v target]
  (let [initial-state (->State 0 0 x-v y-v)]
    (loop [n 0 state initial-state y-max Integer/MIN_VALUE]
      (cond
        (> n 1000) nil
        (overshot? target state) nil
        (hit-target? target state) y-max
        :else (let [new-state (update-state state)]
                (recur (inc n) new-state (max y-max (:y new-state))))))))

(defn -main []
  (let [target (->Target 85 145 -163 -108)
        all-heights (for [x-v (range -200 200)
                          y-v (range -200 200)]
                   (run-simulation x-v y-v target))
        heights (remove nil? all-heights)
        part1 (apply max heights)
        part2 (count heights)]
  (run! println [part1 part2])))
